core/src/main/java/org/mule/runtime/core/internal/streaming/FunkyCursorManager.java

```diff
@@ -0,0 +1,239 @@
+
+  private final LoadingCache<StreamingSessionKey, EventStreamingState> registry =
+      Caffeine.newBuilder()
+          .removalListener((key, value, cause) -> ((EventStreamingState) value).dispose())
+          .build(key -> {
+            hookEventTermination(key);
+            return new EventStreamingState();
+          });
+
+  private final MutableStreamingStatistics statistics;
+  private final StreamingGarbageCollector streamingGarbageCollector;
+
+  /**
+   * Creates a new instance
+   *
+   * @param statistics statistics which values should be kept updated
+   */
+  public FunkyCursorManager(MutableStreamingStatistics statistics, Scheduler scheduler) {
+    super(statistics);
+    this.statistics = statistics;
+    streamingGarbageCollector = new StreamingGarbageCollector(scheduler);
+  }
+
+  @Override
+  public void start() throws MuleException {
+    streamingGarbageCollector.start();
+  }
+
+  @Override
+  public void stop() throws MuleException {
+    streamingGarbageCollector.stop();
+  }
+
+  /**
+   * Becomes aware of the given {@code provider} and returns a replacement provider which is managed by the runtime, allowing for
+   * automatic resource handling
+   *
+   * @param provider     the provider to be tracked
+   * @param ownerContext the root context of the event that created the provider
+   * @return a {@link CursorProvider}
+   */
+  @Override
+  public CursorProvider manage(CursorProvider provider, BaseEventContext ownerContext) {
+    final CursorContext context = new CursorContext(provider, ownerContext);
+
+    ManagedCursorProvider managedProvider;
+    if (provider instanceof CursorStreamProvider) {
+      managedProvider = new ManagedCursorStreamProvider(context, this, statistics);
+    } else if (provider instanceof CursorIteratorProvider) {
+      managedProvider = new ManagedCursorIteratorProvider(context, this, statistics);
+    } else {
+      throw new MuleRuntimeException(createStaticMessage("Unknown cursor provider type: " + context.getClass().getName()));
+    }
+
+    registry.get(new StreamingSessionKey(ownerContext)).addProvider(managedProvider);
+
+    return managedProvider;
+  }
+
+  /**
+   * Acknowledges that the given {@code cursor} has been opened
+   *
+   * @param cursor         the opened cursor
+   * @param providerHandle the handle for the provider that generated it
+   */
+  @Override
+  public void onOpen(Cursor cursor, CursorContext providerHandle) {
+    //registry.get(providerHandle.getOwnerContext().getId()).addCursor(providerHandle.getCursorProvider(), cursor);
+    //statistics.incrementOpenCursors();
+  }
+
+
+  /**
+   * Acknowledges that the given {@code cursor} has been closed
+   *
+   * @param cursor the closed cursor
+   * @param handle the handle for the provider that generated it
+   */
+  @Override
+  public void onClose(Cursor cursor, CursorContext handle) {
+    //final String eventId = handle.getOwnerContext().getId();
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: This block of commented-out lines of code should be removed. ([squid:CommentedOutCodeLine](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7643&issues=AWirZZoQotIXnmsUDp_i&open=AWirZZoQotIXnmsUDp_i)

[ISSUE_KEY]: AWirZZoQotIXnmsUDp_i